Cloudpresser/user details tooltip

cloudpresser opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
- Create a ***UserDetailsTooltip*** component, which uses 2 prop, `accountID, and fallbackUserDetails`
- ***UserDetailsTooltip*** returns a ***Tooltip*** component with a custom renderTooltipContent and passes it's children into it, like this:
```js
const UserDetailsTooltip = props => {
    const { avatarSource, name, handle, children } = props

    return (
        <Tooltip
            renderTooltipContent={()=> (
                    <View>
                        <Avatar
                            containerStyles={[styles.actionAvatar]}
                            source={avatarSource}
                        />
                        <Text>{name}</Text>
                        <Text>@{handle}</Text>
                    </View>
                )}
            children={children}
        />
    )
}
```
- Modified call sites to use ReportUtils.getPersonalDetailsForLogin to get userDetails object for login and access accountID.
- Used fallback object to handle cases where personalDetails are not in Onyx (app@t.com is one such case).

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/20086
$ https://github.com/Expensify/App/issues/20086#issuecomment-1574538614


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console
- [x] Run app on web
- [x] Hover over the user name or image on a chat, sidebar, or anywhere user tooltips were previously used, except mentions
- [x] Verify new tooltip style is being used
- [x] Verify component renders well with and without a login, name and avatar

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
- [x] Native and mobile platforms do not apply, verify app does not have regressions (no-op)
- [x] Verify that no errors appear in the JS console
- [x] Run app on web
- [x] Hover over the user name or image on a chat, sidebar, or anywhere user tooltips were previously used, except mentions
- [x] Verify new tooltip style is being used
- [x] Verify component renders well with and without a login, name and avatar
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
- [x] Native and mobile platforms do not apply, verify app does not have regressions (no-op)
- [x] Verify that no errors appear in the JS console
- [x] Run app on web
- [x] Hover over the user name or image on a chat
- [x] Verify new tooltip style is being used

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->
<img width="409" alt="Screenshot 2023-06-11 at 11 16 46 AM" src="https://github.com/Expensify/App/assets/59715908/b521ff4c-263e-4afc-a7fb-218041324ed0">


</details>

<details>
<summary>Mobile Web - Chrome</summary>
Does not apply (component does not render on mobile)
<!-- add screenshots or videos here -->



</details>

<details>
<summary>Mobile Web - Safari</summary>
Does not apply (component does not render on mobile)
<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->
<img width="409" alt="Screenshot 2023-06-11 at 11 18 37 AM" src="https://github.com/Expensify/App/assets/59715908/de406760-9e41-4018-90d6-e4309630ab47">

</details>

<details>
<summary>iOS</summary>
Does not apply (component does not render on mobile)
<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>
Does not apply (component does not render on mobile)
<!-- add screenshots or videos here -->

</details>

cloudpresser commented:
Still waiting for a high-traffic account at luiz+high-traffic-1@cloudpresser.com

cloudpresser commented:
I have read the CLA Document and I hereby sign the CLA

puneetlath commented:
@cloudpresser let's make sure the tooltip is also getting the proper styling. And can you try setting a handle manually in Onyx to show what it looks like with the handle as well. 

<img width="516" alt="Screenshot 2023-06-06 at 11 49 55 AM" src="https://github.com/Expensify/App/assets/1371996/aa02fd9f-703b-4392-839d-1506cd1de7d8">

<img width="506" alt="Screenshot 2023-06-06 at 11 50 08 AM" src="https://github.com/Expensify/App/assets/1371996/cf2e2264-093e-4951-ad7c-ec6f6066d657">

<img width="508" alt="Screenshot 2023-06-06 at 11 50 18 AM" src="https://github.com/Expensify/App/assets/1371996/b7a2c551-0819-409a-9b70-e67aa0eb076e">

cloudpresser committed:
feat(userTooltip): init

cloudpresser committed:
fix(userTooltip): proper boolean conditional

cloudpresser committed:
feat(reportActionItem): use userTooltip

cloudpresser committed:
fix(userDetailsTooltip): update imports

cloudpresser committed:
fix(userDetailsTooltip): rename index

fedirjh reviewed (changes requested):
Looks good , left few comments

fedirjh reviewed (changes requested):
An extra suggestion

cloudpresser committed:
style(userDetailsTooltip): use function instead of const

Co-authored-by: Fedi Rajhi <fedirjh@gmail.com>

cloudpresser committed:
style(userDetailsTooltip): cast string to boolean

Co-authored-by: Fedi Rajhi <fedirjh@gmail.com>

cloudpresser committed:
style(userDetailsTooltip): don't destructure props

cloudpresser committed:
fix(userDetailsTooltip): useCallback

cloudpresser committed:
style(userDetailsTooltip): new line

cloudpresser commented:
> @cloudpresser let's make sure the tooltip is also getting the proper styling. And can you try setting a handle manually in Onyx to show what it looks like with the handle as well.
> 
> <img alt="Screenshot 2023-06-06 at 11 49 55 AM" width="516" src="https://user-images.githubusercontent.com/1371996/243754153-aa02fd9f-703b-4392-839d-1506cd1de7d8.png"> <img alt="Screenshot 2023-06-06 at 11 50 08 AM" width="506" src="https://user-images.githubusercontent.com/1371996/243754208-cf2e2264-093e-4951-ad7c-ec6f6066d657.png"> <img alt="Screenshot 2023-06-06 at 11 50 18 AM" width="508" src="https://user-images.githubusercontent.com/1371996/243754273-b7a2c551-0819-409a-9b70-e67aa0eb076e.png">

In order for to handle the case with no handle or name, the dimensions of the overall tooltip should probably use paddings instead of a fixed height. Looking at the screenshots I can only guess the value for those margins, could you tell me: 
- What is the distance from ***top of avatar*** to the ***top of the tooltip***
- What is the distance from ***bottom of handle*** to the ***bottom of the tooltip***

Alternatively I could extract it from the design files, but I'm not sure where or if I can access those.

cloudpresser commented:
> @cloudpresser For style consistency , we can use same style used in [ReactionTooltipContent.js](https://github.com/Expensify/App/blob/5739db4953e1546ba2aa113d0a16d26bfc0c8898/src/components/Reactions/ReactionTooltipContent.js#L34)
> 
> https://github.com/Expensify/App/blob/5739db4953e1546ba2aa113d0a16d26bfc0c8898/src/components/Reactions/ReactionTooltipContent.js#L44-L59

Are you suggesting I use this for the name? It looks like these styles do not match the ones on the screenshot @puneetlath shared. I was looking at `styles.h4` for title and `styles.textMicroMuted` (to be created) for handle
```
textMicroMuted: {
        color: themeColors.greenIcons,
        fontFamily: fontFamily.EXP_NEUE,
        fontSize: variables.fontSizeSmall,
        lineHeight: variables.lineHeightSmall,
    },
```

cloudpresser committed:
fix(userDetailsTooltip): propTypes

cloudpresser commented:
> We have a console warning as well
> 
> <img alt="Screenshot 2023-06-06 at 7 38 23 PM" width="1439" src="https://user-images.githubusercontent.com/36869046/243795231-8292d88d-53b5-48a6-ba35-6a7c83b0b955.png">

Just resolved this. Sorry, that got lost in a stash earlier

cloudpresser commented:
> > Are you suggesting I use this for the name?
> 
> I am suggesting to use same template , same style classes , change emoji code with avatar , emoji name with handle ...

That makes sense, and saves me from guessing on those paddings. Thank you for the suggestion and for clarifying

cloudpresser committed:
feat(userDetailsTooltip): correct styles

fedirjh reviewed (changes requested):
@cloudpresser We should update all tooltips instances that display the user's email address

Examples :

- LHN 
<img width="378" alt="LHN" src="https://github.com/Expensify/App/assets/36869046/8ef6bac3-0168-4665-9c99-6486ac764bdd">

- Search Page

<img width="490" alt="Search" src="https://github.com/Expensify/App/assets/36869046/312f6dde-1964-4eda-8108-98a830e45dc3">

- Chat Header

<img width="490" alt="Header" src="https://github.com/Expensify/App/assets/36869046/a9eca930-1d2d-458a-8da8-fd8e57f7b23a">

- Details Page

<img width="490" alt="details" src="https://github.com/Expensify/App/assets/36869046/312c1a74-15d2-417a-bd5d-581a3067a904">

- Threads

<img width="490" alt="Threads" src="https://github.com/Expensify/App/assets/36869046/2a6e8f1f-ec11-4fcf-b2e5-f532dc1fd26d">

- Mentions

<img width="490" alt="Mentions" src="https://github.com/Expensify/App/assets/36869046/d1656de1-f590-4b15-8a55-229f8742fb79">

cloudpresser commented:
@fedirjh  I have not been granted a high traffic account yet. My understanding is I should use that to test cases like the one you found a bug in, is that right? I requested it twice so far on slack `luiz+high-traffic-1@cloudpresser.com`

puneetlath commented:
@cloudpresser just made luiz+high-traffic-1@cloudpresser.com a high traffic account.

cloudpresser committed:
feat(userDetailsTooltip): use login instead of handle

cloudpresser commented:
Still working on this. The tooltips I had not accounted for are a bit of a deeper hole than anticipated. Will need to come back to it after lunch time. Expect another update before 5pm EST.

cloudpresser committed:
feat(userDetailsTooltip): use on LHN users

cloudpresser commented:
Getting in a groove but my time is out for today. I'll be back tomorrow, expect more updates before 12pm. Realistically, I may have addressed all remaining by 5m EST tomorrow.

cloudpresser committed:
chore: lint

cloudpresser committed:
Merge remote-tracking branch 'upstream/main' into cloudpresser/userDetailsTooltip

cloudpresser committed:
feat(userDetailsTooltip): use it in mentions

cloudpresser committed:
feat(userDetailsTooltip): use it in detailsPage

puneetlath commented:
How are things going here?

puneetlath commented:
I think variable width is fine, given that email addresses can also be variable. I believe that's how they are today also.

cloudpresser commented:
Hey @puneetlath @fedirjh, I am still working on this. Today I was able to implement the tooltip on all the highlighted locations, but have not yet fixed the bug as @fedirjh mentioned [here](https://github.com/Expensify/App/pull/20276#issuecomment-1583461329).

Tomorrow, I am planning on focusing on [this bug](https://github.com/Expensify/App/pull/20276#issuecomment-1579318630) as I have not yet found the source of it, and maybe implement @fedirjh's [suggestion](https://github.com/Expensify/App/pull/20276#issuecomment-1583536673) because it does seem very good and something I thought of but haven't had enough context to come up with yet.

@fedirjh I'll also be rebasing these commits tomorrow to make sure they're signed.

It really does look to me like we'll be able to close this tomorrow. I'll have the day off on my job too, so more time to focus on this than the last few days.

puneetlath commented:
Ok thanks for the updates!

>@puneetlath @cloudpresser I have another suggestion that we can consider. Instead of passing all props down to the UserDetailsTooltip component, we can modify it to only accept the email (login) as a prop. Then, we can connect the UserDetailsTooltip component to Onyx.personalDetails and use the email to retrieve data from personalDetails. I believe implementing this change will enhance the maintainability of the code.

This plan sounds good, however we should use the accountID instead of the email address and use the `personalDetailsList` object in Onyx instead of `personalDetails`. We are in the process of migrating from `personalDetails` to `personalDetailsList` which is keyed by accountIDs instead of emails. `personalDetailsList` will contain different data depending on the user. Sometimes it might include their login/displayName/etc. Sometimes it might just have their avatar. 

So that would actually work well here. The only case where it wouldn't work is for mentions, which don't currently have an accountID associated with them. But I'm fine with just leaving mentions to use the old tooltip for now.

puneetlath commented:
I think it's not worth it. Mentions will be updated in the near-ish future to be accountID-based. For now I think it's fine to leave them using the old tooltip.

puneetlath commented:
@cloudpresser still on track to have this ready for final review today?

puneetlath commented:
And I just want to be sure, the changes in this PR gracefully handle the scenario where any of these fields aren't set: displayName, login.

cloudpresser committed:
Revert "feat(userDetailsTooltip): use it in mentions"

This reverts commit c8f6f45425745e19108797402a9e3fb0941e90cf.

cloudpresser commented:
Today is not my day. I'll work on this over the weekend.

puneetlath commented:
Ok sounds good. I'll also be working on the weekend, so ping me when you're ready for more review.

puneetlath commented:
I'm also happy to extend the 50% bonus timeline if we can get this merged by Monday.

cloudpresser committed:
refactor(userDetailsTooltip): use accountID

cloudpresser committed:
feat(userDetailsTooltip): use it on report welcome

cloudpresser committed:
fix(userDetailsTooltip): render issue

cloudpresser commented:
> 

Sounds good, I'm already working to make that possible

cloudpresser committed:
fix(sideBarLinks): pass accountID to avatar

cloudpresser committed:
fix(sidebarLinks): revert tooltip changes

cloudpresser commented:
@puneetlath I think I'm ready for further review. I completed:
- [x] Fixed [bug](https://github.com/Expensify/App/pull/20276#issuecomment-1579318630)
- [x] [Refactored](https://github.com/Expensify/App/pull/20276#issuecomment-1583536673) to use [accountID](https://github.com/Expensify/App/pull/20276#issuecomment-1583578426)
- [x] Implemented in details page
- [x] [Reverted changes mentions](https://github.com/Expensify/App/pull/20276#issuecomment-1583611120), as they are pending refactor
- [x] [Rebased commits to be signed](https://github.com/Expensify/App/pull/20276#issuecomment-1583462157)
- [x] Fixed other small issues

puneetlath commented on src/components/DisplayNames/displayNamesPropTypes.js:
```diff
@@ -11,7 +11,10 @@ const propTypes = {
-            tooltip: PropTypes.string,
+            tooltip: PropTypes.shape({
+               /** The Account Id for the tooltip */ 
+                accountId: PropTypes.string,
```
```suggestion
                accountID: PropTypes.string,
```

puneetlath commented on src/components/DisplayNames/displayNamesPropTypes.js:
```diff
@@ -11,7 +11,10 @@ const propTypes = {
             /** The tooltip to show when the associated name is hovered */
-            tooltip: PropTypes.string,
+            tooltip: PropTypes.shape({
+               /** The Account Id for the tooltip */ 
```
```suggestion
               /** The Account ID for the tooltip */ 
```

puneetlath commented on src/components/AvatarWithIndicator.js:
```diff
@@ -22,9 +22,9 @@ import themeColors from '../styles/themes/default';
-    tooltipText: PropTypes.string,
+    
+     /** To show a tooltip on hover */
+     tooltipText: PropTypes.string,
```
I believe this should be replaced by accountID

cloudpresser replied:
This component seems to be poorly named. I thought the same at first, but this avatar is only used on the sidebar header, where it links to settings. So I think a tooltipText is more appropriate here (or it really will not be clear it takes you to settings)
<img width="250" alt="Screenshot 2023-06-10 at 4 32 44 PM" src="https://github.com/Expensify/App/assets/59715908/d2009750-e73b-40a0-8272-f6905ad2948a">

cloudpresser replied:
This is only use of this component 

https://github.com/Expensify/App/blob/035783a7984e0214677b5364d9351c968887bf58/src/pages/home/sidebar/SidebarLinks.js#LL202C33-L205C35

puneetlath commented on src/libs/ReportUtils.js:
```diff
@@ -2154,6 +2157,7 @@ function getParentReport(report) {
 
 export {
     getReportParticipantsTitle,
+    getPersonalDetailsForLogin,
```
What is this for?

cloudpresser replied:
I used in in tooltips where only the login was available to get the accountID. Is there a better alternative?

puneetlath commented on src/pages/home/report/ReportActionItemSingle.js:
```diff
@@ -67,7 +67,7 @@ const showUserDetails = (accountID) => {
 const ReportActionItemSingle = (props) => {
     const actorEmail = props.action.actorEmail.replace(CONST.REGEX.MERGED_ACCOUNT_PREFIX, '');
-    const {accountID, avatar, displayName, pendingFields} = props.personalDetails[actorEmail] || {};
+    const {accountID, avatar, displayName, pendingFields, login} = props.personalDetails[actorEmail] || {};
```
What does `login` get used for?

cloudpresser replied:
It is not, I missed this in the refactor. Reverting now

puneetlath commented on src/components/UserDetailsTooltip/userDetailsTooltipPropTypes.js:
```diff
@@ -0,0 +1,24 @@
+import PropTypes from 'prop-types';
+
+const propTypes = {
+    /** User's Account Id */
```
```suggestion
    /** User's Account ID */
```

puneetlath commented on src/components/UserDetailsTooltip/userDetailsTooltipPropTypes.js:
```diff
@@ -0,0 +1,24 @@
+const propTypes = {
+    /** User's Account Id */
+    accountID: PropTypes.string.isRequired,
+    /** Fallback User Details object used if no account id */
```
```suggestion
    /** Fallback User Details object used if no accountID */
```

puneetlath commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -19,14 +18,21 @@ import {withNetwork} from '../../../components/OnyxProvider';
+
+    /** User's login */
+    login: PropTypes.string,
+
```
Why did you add these? Did you plan to use them for fallback?

cloudpresser replied:
I missed these in the refactor, fixing it now

puneetlath reviewed (changes requested):
Some mostly minor comments. 

Looks like you have some lint errors and a jest error as well. You can run lint locally with `npm run lint`.

cloudpresser committed:
Update src/components/DisplayNames/displayNamesPropTypes.js

Co-authored-by: Puneet Lath <puneet@expensify.com>

cloudpresser committed:
Update src/components/UserDetailsTooltip/userDetailsTooltipPropTypes.js

Co-authored-by: Puneet Lath <puneet@expensify.com>

cloudpresser committed:
Update src/components/DisplayNames/displayNamesPropTypes.js

Co-authored-by: Puneet Lath <puneet@expensify.com>

cloudpresser committed:
Update src/components/UserDetailsTooltip/userDetailsTooltipPropTypes.js

Co-authored-by: Puneet Lath <puneet@expensify.com>

cloudpresser committed:
fix(reportActionItem): fragment props

cloudpresser committed:
fix(reportActionItemMessage): pass accountID

cloudpresser committed:
chore: eslint --fix

cloudpresser committed:
chore: remove unused prop

fedirjh reviewed (changes requested):
Left few comments , overall looks good , the previous bug was fixed ✅

cloudpresser committed:
fix: fallback avatar not avatarSource

cloudpresser committed:
fix(userDetailsTooltip): avatar propType

cloudpresser commented on src/libs/ReportUtils.js:
```diff
@@ -873,8 +873,11 @@ function getDisplayNameForParticipant(login, shouldUseShortForm = false) {
         const displayName = getDisplayNameForParticipant(participant.login, isMultipleParticipantReport);
-        const tooltip = participant.login ? Str.removeSMSDomain(participant.login) : '';
 
+        const tooltip = {
+            accountID: participant.accountID,
+        }
```
Actually this console error is because avatar souce can in fact be a PropTypes.func, but my proptype incorrectly only accepted strings. Fixed it [here](https://github.com/Expensify/App/pull/20276/commits/8f852403523161a28c62ee7103baadda82b719aa)

cloudpresser commented:
> For some emails (`app@t.com`) the tooltip is not showing , for mentions as well .
> 
>  Screen.Recording.2023-06-10.at.11.10.45.PM.mov 
>  Screen.Recording.2023-06-10.at.11.19.05.PM.mov

Fixed the app@t.com email scenario. It was an issue with the fallback object. Not sure what happened with mentions because we're no longer touching it in the PR. However, it is possible it was impacted by a fallback absence, because they are working on my local branch after the fix.

https://github.com/Expensify/App/assets/59715908/17a9ded2-96c4-4511-90a2-a64527b47cab

cloudpresser commented on src/components/MultipleAvatars.js:
```diff
@@ -173,7 +182,9 @@ const MultipleAvatars = (props) => {
                 <View style={singleAvatarStyles}>
-                    <Tooltip text={tooltipTexts[0]}>
+                    <UserDetailsTooltip
+                        accountID={ReportUtils.getPersonalDetailsForLogin(props.icons[0].name).accountID}
```
It returns a `personalDetails` object for the login. personalDetails has accountID, which I'm accessing here. 
I looked for an util returning just the `accountID`, but I did not find one. Should I create one instead?

cloudpresser committed:
feat(userDetailsTooltip): remove sms domain

cloudpresser committed:
chore: lint

cloudpresser commented on src/libs/ReportUtils.js:
```diff
@@ -873,8 +873,11 @@ function getDisplayNameForParticipant(login, shouldUseShortForm = false) {
 function getDisplayNamesWithTooltips(participants, isMultipleParticipantReport) {
     return _.map(participants, (participant) => {
         const displayName = getDisplayNameForParticipant(participant.login, isMultipleParticipantReport);
-        const tooltip = participant.login ? Str.removeSMSDomain(participant.login) : '';
```
Done

cloudpresser committed:
fix(userDetailsTooltip): condition to hide

cloudpresser committed:
chore: use lodashGet

cloudpresser committed:
chore: use lodashget

cloudpresser commented on src/components/MultipleAvatars.js:
```diff
@@ -74,7 +76,14 @@ const MultipleAvatars = (props) => {
+                    displayName: ReportUtils.getPersonalDetailsForLogin(props.icons[0].name).displayName || '',
+                    login: props.icons[0].name || tooltipTexts[0],
+                    avatarSource: props.icons[0].source || '',
+                }}
```
Done

cloudpresser committed:
test: update expectations

puneetlath commented:
Ok nice! I'll take another look later today. Thanks for the hustle!

fedirjh reviewed (changes requested):
@cloudpresser It look like you made some changes to [src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.js](https://github.com/cloudpresser/New-Expensify/blob/cloudpresser/userDetailsTooltip/src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.js) and that's break the mention tooltips. can you please take a look , here is the local PR branch

<img width="1005" alt="Screenshot 2023-06-11 at 7 18 28 PM" src="https://github.com/Expensify/App/assets/36869046/06657628-0213-494f-8196-f929d36317f0">

cloudpresser commented:
> @cloudpresser It look like you made some changes to [src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.js](https://github.com/cloudpresser/New-Expensify/blob/cloudpresser/userDetailsTooltip/src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.js) and that's break the mention tooltips. can you please take a look , here is the local PR branch
> 
> <img alt="Screenshot 2023-06-11 at 7 18 28 PM" width="1005" src="https://user-images.githubusercontent.com/36869046/244958045-06657628-0213-494f-8196-f929d36317f0.png">

That doesn't match my current diff, I changed it back in [Revert "feat(userDetailsTooltip): use it in mentions"](https://github.com/Expensify/App/pull/20276/commits/efd0580a12549aebdd31a265d873003f106bc104)

cloudpresser commented on src/components/UserDetailsTooltip/UserDetailsTooltip.js:
```diff
@@ -0,0 +1,48 @@
+      , [userDetails.avatar, userDetails.displayName, userDetails.login]);
+
+    return (
+        <Tooltip renderTooltipContent={(userDetails.name || userDetails.login) ? renderTooltipContent : undefined}>
```
We can. Changing it...

cloudpresser replied:
Actually, if I do it this way it may require changes to the tooltip. A callback that returns null is still evaluated as non-null for the purpose of determining if the renderTooltipContent should be used. This causes tooltips to render with empty content. 

https://github.com/Expensify/App/blob/035783a7984e0214677b5364d9351c968887bf58/src/components/Tooltip/TooltipRenderedOnPageBody.js#LL118C5-L120C7

cloudpresser replied:
I'm not sure this would be the best idea, because it changes behavior of the underlying Tooltip component, and I cannot be sure what code relies on that behavior.

cloudpresser replied:
Only when accountID does not map to a user and no fallback is provided (one such example is the concierge chat)

cloudpresser replied:
> I think we can just make an early `return` , but do we have any use case where `login` is missing ?

I think that would work, I'll try it now

cloudpresser replied:
Done

cloudpresser replied:
Yes, that was the original behavior and I thought it was outside the scope of this issue to fix that. There are a few more places that did not display the tooltip originally where it looks like it should. I think if we start digging here the scope of these changes will increase significantly.

cloudpresser replied:
> Hmm, I believe the concierge has an email (login), and in my opinion, we should always show the tooltip for the concierge. However, for the chat header, the tooltip is not shown for the concierge.
> 
>  Screen.Recording.2023-06-11.at.10.31.44.PM.mov

Just fixed it, had to modify ReportUtils.getIcons to return concierge email on name for tooltip.

cloudpresser commented on src/components/UserDetailsTooltip/UserDetailsTooltip.js:
Ok

cloudpresser committed:
chore: rename UserDetailsTooltip.js to index.js

cloudpresser committed:
chore: prettier

cloudpresser committed:
feat(userDetailsTooltip): early return

cloudpresser committed:
fix: add concierge email on name for tooltip

puneetlath commented:
@fedirjh let's go ahead and file a separate bug for that. I have a vague memory of that being raised and it may even be fixed already.

puneetlath commented:
How are things looking in general to you @fedirjh? Do you feel we are close to merge? 

@cloudpresser I see we still have a lint error. Looks like you need to run `npm run prettier`

cloudpresser committed:
chore: prettier

puneetlath commented:
>do you think we need any style enhancements or the actual result is good ?

Ah, I see that much earlier there was a question about padding. In figma we have 16px on the sides and 12px on top/bottom of the tooltip. And then 8px between avatar and name. Otherwise style looks good to me.

cloudpresser committed:
fix: styles

cloudpresser committed:
refactor: create getAccoutIdForLogin function

cloudpresser commented on src/components/DisplayNames/displayNamesPropTypes.js:
```diff
@@ -11,7 +11,10 @@ const propTypes = {
 
             /** The tooltip to show when the associated name is hovered */
-            tooltip: PropTypes.string,
+            tooltip: PropTypes.shape({
+                /** The Account ID for the tooltip */
+                accountID: PropTypes.string,
+            }),
```
I think we can, pushing a commit.

cloudpresser replied:
Yes it was, updated it now.

cloudpresser commented on src/components/MultipleAvatars.js:
```diff
@@ -185,10 +195,10 @@ const MultipleAvatars = (props) => {
                     <View style={secondAvatarStyles}>
                         {props.icons.length === 2 ? (
-                            <Tooltip text={tooltipTexts[1]}>
+                            <UserDetailsTooltip accountID={lodashGet(ReportUtils.getPersonalDetailsForLogin(props.icons[1].name), 'accountID', '')}>
```
Sure

cloudpresser committed:
test: update expectations

cloudpresser committed:
feat(getDisplayNamesWithTooltips): use accountID directly

puneetlath commented:
@cloudpresser prettier error again.

cloudpresser committed:
chore: prettier

cloudpresser commented:
Yes, sorry I am very used to doing prettier through eslint. Adjusting my settings...

cloudpresser committed:
fix: displayNamesPropTypes accountID

fedirjh reviewed (changes requested):
@puneetlath Code looks good to me. there is one little case , if the user remove their names , the name will fallback to their email , and that will cause the tooltip to display email twice. 

<img width="341" alt="twice" src="https://github.com/Expensify/App/assets/36869046/2627d30d-cf01-4ad3-9a65-e7c9d845e29a">

cloudpresser committed:
fix: accept numbers or strings

Co-authored-by: Fedi Rajhi <fedirjh@gmail.com>

cloudpresser committed:
Update src/components/MultipleAvatars.js

Co-authored-by: Fedi Rajhi <fedirjh@gmail.com>

cloudpresser committed:
feat: hide email when displayName is equal

cloudpresser commented:
> @puneetlath Code looks good to me. there is one little case , if the user remove their names , the name will fallback to their email , and that will cause the tooltip to display email twice.
> 
> <img alt="twice" width="341" src="https://user-images.githubusercontent.com/36869046/244982627-2627d30d-cf01-4ad3-9a65-e7c9d845e29a.png">

I can check if they're the same and hide the greyed out email. I'm committing it now but can reverse if that's not expected behavior. 

Like this:
<img width="380" alt="Screenshot 2023-06-11 at 8 45 35 PM" src="https://github.com/Expensify/App/assets/59715908/4290acd9-f473-45a9-9f2d-f0ea01d9478d">

puneetlath commented:
Sounds good. I like that.

cloudpresser committed:
fix: do not return text inside view

Co-authored-by: Fedi Rajhi <fedirjh@gmail.com>

fedirjh reviewed (changes requested):
@cloudpresser tooltips looks good and works well for most case , however I found some cases where tooltips doesn’t show , for example if we search for a multi participants conversation let's say 8 , the search result doesn’t have tooltip , IT doesn’t show on header as well , I think we should add a `fallbackUserDetails` to `DisplayNames` , we should update `ReportUtils.getDisplayNamesWithTooltips` to return avatar + login along with `displayName` and `accountID` , Note that the tests are with email that doesn’t have an accountID:

### Multiple emails:

<img width="371" alt="Screenshot 2023-06-12 at 12 46 21 PM" src="https://github.com/Expensify/App/assets/36869046/cc277253-851a-4053-9d76-131a42930f34">

### Multi Participants Report Header

https://github.com/Expensify/App/assets/36869046/1ba2c11c-ead5-4d60-b727-76bef655aa27

### Search result 

| Staging | Dev |
|--------|--------|
| <video src=https://github.com/Expensify/App/assets/36869046/0a342785-527f-469b-bdeb-780075b4cab1 /> | <video src=https://github.com/Expensify/App/assets/36869046/7fc253a8-9d52-4091-a959-25cbe979762d /> | 

### Details Page

https://github.com/Expensify/App/assets/36869046/1a7726b0-11ac-4815-90cb-6a914b23fecf

cloudpresser committed:
Update src/components/UserDetailsTooltip/index.js

Co-authored-by: Fedi Rajhi <fedirjh@gmail.com>

cloudpresser commented:
Updated

> @cloudpresser tooltips looks good and works well for most case , however I found some cases where tooltips doesn’t show , for example if we search for a multi participants conversation let's say 8 , the search result doesn’t have tooltip , IT doesn’t show on header as well , I think we should add a `fallbackUserDetails` to `DisplayNames` , we should update `ReportUtils.getDisplayNamesWithTooltips` to return avatar + login along with `displayName` and `accountID` , Note that the tests are with email that doesn’t have an accountID:
> 
> ### Multiple emails:
> <img alt="Screenshot 2023-06-12 at 12 46 21 PM" width="371" src="https://user-images.githubusercontent.com/36869046/245156173-cc277253-851a-4053-9d76-131a42930f34.png">
> ### Multi Participants Report Header
>  Screen.Recording.2023-06-12.at.2.17.22.PM.mov 
> ### Search result
> Staging	Dev
>  Screen.Recording.2023-06-12.at.2.02.01.PM.mov 
>  Screen.Recording.2023-06-12.at.2.42.34.AM.mov 
> ### Details Page
>  Screen.Recording.2023-06-12.at.2.45.39.PM.mov

Modifying `ReportUtils.getDisplayNamesWithTooltips` works!

<img width="998" alt="Screenshot 2023-06-12 at 10 06 03 AM" src="https://github.com/Expensify/App/assets/59715908/1b8b502f-ef95-463c-9d6c-6be00884201c">
<img width="373" alt="Screenshot 2023-06-12 at 10 14 29 AM" src="https://github.com/Expensify/App/assets/59715908/153755e7-62f6-4eeb-bfee-c0e17d74ee84">

cloudpresser committed:
test: update displayNamesWithTooltips expectation

cloudpresser committed:
feat(displayNamesWithTooltips): avatar and login

cloudpresser committed:
feat(displayNames): fallback tooltips

cloudpresser commented:
Updated

cloudpresser commented:
### Search 8 participants with displayName and email
<img width="373" alt="Screenshot 2023-06-12 at 10 35 23 AM" src="https://github.com/Expensify/App/assets/59715908/33479fc1-18f2-40ef-95cc-ea5dee407fa1">

### Header 8 participants with displayName and email
<img width="373" alt="Screenshot 2023-06-12 at 10 36 19 AM" src="https://github.com/Expensify/App/assets/59715908/7474fa38-fc1b-44ac-909e-4f141ee57166">

cloudpresser commented:
### Details page

https://github.com/Expensify/App/assets/59715908/34f11a0f-5754-45aa-8834-a80390d48ae2

fedirjh reviewed (approved):
Thanks for the great work @cloudpresser , It looks good and tests well for all cases 🚀 🚀 

All yours @puneetlath

puneetlath commented:
Thank you both so much for your persistence and urgency on this one! It's really much appreciated. Looking forward to more contributions from you @cloudpresser.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5249660781) to staging by https://github.com/puneetlath in version: 1.3.27-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5249660781) to staging by https://github.com/puneetlath in version: 1.3.27-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

bondydaa commented on src/libs/ReportUtils.js:
```diff
@@ -873,7 +884,8 @@ function getDisplayNameForParticipant(login, shouldUseShortForm = false) {
     return _.map(participants, (participant) => {
         const displayName = getDisplayNameForParticipant(participant.login, isMultipleParticipantReport);
-        const tooltip = participant.login ? Str.removeSMSDomain(participant.login) : '';
+        const avatar = UserUtils.getDefaultAvatar(participant.login);
```
I think this caused a regression where we are unable to load a user's uploaded/custom avatar in some cases as identified on https://github.com/Expensify/App/issues/20683

I believe I've got a fix so will push it up and request reviews from ya'll.

bondydaa replied:
PR is here https://github.com/Expensify/App/pull/20727

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5272016379) to production by https://github.com/AndrewGable in version: 1.3.27-7 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

grgia commented:
I just reported this, but it seems we didn't handle the case for Workspace Avatars https://expensify.slack.com/archives/C049HHMV9SM/p1686821040533159

esh-g commented on src/libs/ReportUtils.js:
```diff
@@ -845,6 +846,16 @@ function getPersonalDetailsForLogin(login) {
+ * @param {String} login
+ * @returns {String}
+ */
+function getAccountIDForLogin(login) {
```
@cloudpresser @puneetlath
This method no longer works after `accountID` migration because we have `accountID` as they and not `login` anymore. This is causing the following issues: 
- https://github.com/Expensify/App/pull/20746#issuecomment-1595174744
- https://github.com/Expensify/App/pull/20746#issuecomment-1595188280

esh-g replied:
Issue reported here: https://github.com/Expensify/App/issues/20934

Santhosh-Sellavel commented on src/components/UserDetailsTooltip/index.js:
```diff
@@ -0,0 +1,52 @@
+                    {String(userDetails.displayName).trim() ? userDetails.displayName : ''}
+                </Text>
+
+                <Text style={[styles.textMicro, styles.fontColorReactionLabel]}>
+                    {String(userDetails.login).trim() && !_.isEqual(userDetails.login, userDetails.displayName) ? Str.removeSMSDomain(userDetails.login) : ''}
+                </Text>
```
The user with mobile number login case was handled correctly resulting in the [regression](https://github.com/Expensify/App/issues/21197) showing the phone number twice with different formatting. Handled it by improving the checks!

Santhosh-Sellavel commented:
This PR adds UserTooltip and while testing we failed to catch [inconsistency](https://github.com/Expensify/App/pull/20276) in the user tooltip due to a missing account id from the new task page.
